src/main/java/net/minecraftforge/common/ForgeMod.java

```diff
@@ -381,33 +385,37 @@ public ForgeMod(FMLJavaModLoadingContext context) {
         CrashReportCallables.registerCrashCallable("FML", ForgeVersion::getSpec);
         CrashReportCallables.registerCrashCallable("Forge", ()->ForgeVersion.getGroup()+":"+ForgeVersion.getVersion());
 
-        final IEventBus modEventBus = context.getModEventBus();
+        // Todo: [Forge][ForgeMod] mod bus registrations
+        final BusGroup modEventBus = context.getModEventBus();
         // Forge-provided datapack registries
-        modEventBus.addListener((DataPackRegistryEvent.NewRegistry event) -> {
+        // Hacky code, do not copy!
+        EventBus.create(modEventBus, DataPackRegistryEvent.NewRegistry.class).addListener(event -> {
             event.dataPackRegistry(ForgeRegistries.Keys.BIOME_MODIFIERS, BiomeModifier.DIRECT_CODEC);
             event.dataPackRegistry(ForgeRegistries.Keys.STRUCTURE_MODIFIERS, StructureModifier.DIRECT_CODEC);
         });
-        modEventBus.addListener(this::preInit);
-        modEventBus.addListener(this::gatherData);
-        modEventBus.addListener(this::registerFluids);
-        modEventBus.addListener(this::registerVanillaDisplayContexts);
-        modEventBus.addListener(this::onRegisterAttributes);
+        synchronized (ForgeMod.class.getClassLoader()) {
```

Comment:
So this is a major concen that permiates to the whole EventBus.create method. as class loading, and static initalizers are multi threaded. The create method itself needs to be synchronized not the consumers

Reply:
Yeah, it's an intermittent recursive update exception being thrown by ConcurrentHashMap. Need to look into it more and figure out the proper fix

Reply:
Ya, fundamentally I want to has as little effort on the modder's side as possible. And having to synchronize their registrations especially on an object that has nothing to do with events, is too much effort.
So this is a need to fix before being able to pull in.
Tho this sounds like the reason the whole Caching system in EventBus existed.